File: Testing/testing.cpp
#include <iostream>
#include <vector>

#define type int // choose your function type
#define arg_type int // choose your arguments type

class Test{

    private:

        type (*func_)(arg_type) = NULL;
        std::vector <arg_type> params_ = {};
        std::vector <type> results_ = {};

    public:

        void set_func(type (*func)(arg_type)){
            func_ = func;
        }

        bool test(type (*func)(arg_type), arg_type param, type result){
            return func(param) == result;
        }

        bool test(arg_type param, type result){
            if(func_ == NULL){
                std::cout << "No function set" << std::endl;
                return false;
            }
            return func_(param) == result;
        }

        void add_test_case(arg_type param, type result){
            params_.push_back(param);
            results_.push_back(result);
        }

        void show_test_cases(){
            if(params_.size() == 0){
                std::cout << "No test cases" << std::endl;
            }
            for(int i = 0; i < params_.size(); i ++){
                std::cout << "Arguments: " << params_[i] << ", Correct result: " << results_[i] << std::endl;
            }
        }

        void clear_test_cases(){
            params_.clear();
            results_.clear();
        }

        bool test_all(){
            if(func_ == NULL){
                std::cout << "No function set" << std::endl;
                return false;
            }
            for(int i = 0; i < params_.size(); i ++){
                if(func_(params_[i]) != results_[i]){
                    return false;
                }
            }
            return true;
        }

        std::vector <std::pair <arg_type, type> > find_valid(){
            if(func_ == NULL){
                std::cout << "No function set" << std::endl;
                return std::vector <std::pair <arg_type, type> > {{}};
            }
            std::vector <std::pair <arg_type, type> > result = {};
            for(int i = 0; i < params_.size(); i ++){
                try{
                    if(func_(params_[i]) == results_[i]){
                        result.push_back(std::pair <arg_type, type> {params_[i], results_[i]});
                    }
                }
                catch(...){};
            }
            return result;
        }

        std::vector <std::pair <arg_type, type> > find_invalid(){
            if(func_ == NULL){
                std::cout << "No function set" << std::endl;
                return std::vector <std::pair <arg_type, type> > {{}};
            }
            std::vector <std::pair <arg_type, type> > result = {};
            for(int i = 0; i < params_.size(); i ++){
                try{
                    if(func_(params_[i]) != results_[i]){
                        result.push_back(std::pair <arg_type, type> {params_[i], results_[i]});
                    }
                }
                catch(...){}
            }
            return result;
        }

        std::vector <std::pair <arg_type, type> > find_failed(){
            if(func_ == NULL){
                std::cout << "No function set" << std::endl;
                return std::vector <std::pair <arg_type, type> > {{}};
            }
            std::vector <std::pair <arg_type, type> > result = {};
            for(int i = 0; i < params_.size(); i ++){
                try{
                    func_(params_[i]);
                }
                catch(...){
                    result.push_back(std::pair <arg_type, type> {params_[i], results_[i]});
                }
            }
            return result;
        }

        void print_results(){
            if(func_ == NULL){
                std::cout << "No function set" << std::endl;
                return;
            }
            int totaly_valid = 0;
            int totaly_invalid = 0;
            int totaly_failed = 0;
            for(int i = 0; i < params_.size(); i ++){
                std::cout << "Arguments: " << params_[i] << ", Correct result: " << results_[i] << ", Desicion: ";
                try{
                    if(func_(params_[i]) != results_[i]){
                        std::cout << "INVALID" << std::endl;
                        totaly_invalid ++;
                    }
                    else{
                        std::cout << "VALID" << std::endl;
                        totaly_valid ++;
                    }
                }
                catch(...){
                    std::cout << "FAILED" << std::endl;
                    totaly_failed ++;
                }
            }
            std::cout << "Totaly: " << totaly_valid << " Valid, " << totaly_invalid << " INVALID" << totaly_failed << " FAILED" << std::endl;
        }
};

File: Testing/testing.cs
using System;
using T = System.Int32; // choose your function type 
using TArg = System.Int32; // choose your arguments type

namespace Base
{
    
    class Test
    {
        private Func<TArg, T> _func = null;
        private List<TArg> _params = new List<TArg>(){};
        private List<T> _results = new List<T>(){}; 
    
        public void SetFunc(Func<TArg, T> func)
        {
            _func = func;
        }

        public string GetFunc()
        {
            return _func.Method.ToString();
        }

        public bool TestCase(Func<TArg, T> func, TArg param, T result)
        {
            return func(param) == result;
        }

        public bool TestCase(TArg param, T result)
        {
            if(_func == null)
            {
                Console.WriteLine("No function set");
                return false;
            }
            return _func(param) == result;
        }

        public void AddTestCase(TArg param, T result)
        {
            _params.Add(param);
            _results.Add(result);
        }

        public void ShowTestCases()
        {
            if(_params.Count() == 0)
            {
                Console.WriteLine("No test cases");
            }
            for(int i = 0; i < _params.Count(); i ++)
            {
                Console.Write("Arguments: ");
                Console.Write(_params[i]);
                Console.Write(", Correct result: ");
                Console.Write(_results[i]);
                Console.Write("\n");
            }
        }

        public void ClearTestCases()
        {
            _params.Clear();
            _results.Clear();
        }

        public bool TestAll()
        {
            if(_func == null)
            {
                Console.WriteLine("No function set");
                return false;
            }
            for(int i = 0; i < _params.Count(); i ++)
            {
                if(_func(_params[i]) != _results[i])
                {
                    return false;
                }
            }
            return true;
        }

        public List<(TArg param, T result)> FindValid()
        {
            if(_func == null)
            {
                Console.WriteLine("No function set");
                return new List<(TArg param, T result)>(){};
            }
            List<(TArg param, T result)> result = new List<(TArg param, T result)>(){};
            for(int i = 0; i < _params.Count(); i ++)
            {
                try{
                    if(_func(_params[i]) == _results[i])
                    {
                        result.Add((_params[i], _results[i]));
                    }
                }
                catch{};
            }
            return result;
        }

        public List<(TArg param, T result)> FindInvalid()
        {
            if(_func == null)
            {
                Console.WriteLine("No function set");
                return new List<(TArg param, T result)>(){};
            }
            List<(TArg param, T result)> result = new List<(TArg param, T result)>(){};
            for(int i = 0; i < _params.Count(); i ++)
            {
                try{
                    if(_func(_params[i]) != _results[i])
                    {
                        result.Add((_params[i], _results[i]));
                    }
                }
                catch{};
            }
            return result;
        }
    
        public List<(TArg param, T result)> FindFailed()
        {
            if(_func == null)
            {
                Console.WriteLine("No function set");
                return new List<(TArg param, T result)>(){};
            }
            List<(TArg param, T result)> result = new List<(TArg param, T result)>(){};
            for(int i = 0; i < _params.Count(); i ++)
            {
                try{
                    _func(_params[i]);
                }
                catch
                {
                    result.Add((_params[i], _results[i]));
                };
            }
            return result;
        }

        public void PrintResults()
        {
            if(_func == null)
            {
                Console.WriteLine("No function set");
                return;
            }
            Console.Write("Function: ");
            Console.Write(_func.Method);
            Console.Write("\n");
            int totalyValid = 0;
            int totalyInvalid = 0;
            int totalyFailed = 0;
            for(int i = 0; i < _params.Count(); i ++)
            {
                Console.Write("Arguments: ");
                Console.Write(_params[i]);
                Console.Write(", Correct result: ");
                Console.Write(_results[i]);
                Console.Write(", Desicion: ");
                try
                {
                    if(_func(_params[i]) != _results[i])
                    {
                        Console.Write("INVALID");
                        Console.Write("\n");
                        totalyInvalid ++;
                    }
                    else
                    {
                        Console.Write("VALID");
                        Console.Write("\n");
                        totalyValid ++;
                    }
                }
                catch(Exception err)
                {
                    Console.Write("FAILED ( ");
                    Console.Write(err.Message);
                    Console.Write(" ) \n");
                    totalyFailed ++;
                }
            }
            Console.Write("Totaly: ");
            Console.Write(totalyValid);
            Console.Write(" VALID, ");
            Console.Write(totalyInvalid);
            Console.Write(" INVALID, ");
            Console.Write(totalyFailed);
            Console.Write(" FAILED\n");
        }
    }

    class Program
    {
        public static void Main()
        {
            Console.ReadLine();
        }
    }
}